README.md
# bittrex-signalr-client

Node.js implementation of SignalR protocol tailored for Bittrex exchange

## Disclaimer

This is a work in progress mostly meant to be integrated in [Crypto Exchanges Gateway](https://github.com/aloysius-pgast/crypto-exchanges-gateway). But it can also be used as a standalone module to connect to Bittrex WS

NB : doing REST API calls is outside of the scope of the module. If you need support for Bittrex REST API, use [node.bittrex.api](https://github.com/dparlevliet/node.bittrex.api) instead

## What it does

* Implement methods for subscribing to tickers & markets (order books & trades)

* Automatically detect missed order book events and resync automatically by retrieving full order book

* Handle automatic reconnection in (I think !) every possible scenario

* Implements a watchdog to detect when _Bittrex_ stops sending data (default to 30min, automatic reconnection)

* Handle CloudFare's anti-ddos page using [cloudscaper](https://www.npmjs.com/package/cloudscraper/)

* Support for new Bittrex API (tickers, order books, trades & user orders)

## Installation

```
npm install bittrex-signalr-client
```

## How to use it

See [documentation in _doc_ directory](https://github.com/aloysius-pgast/bittrex-signalr-client/tree/master/doc/) for a description of supported API

See [examples in _examples_ directory](https://github.com/aloysius-pgast/bittrex-signalr-client/tree/master/examples/) for an overview of what this library can do

## Other similar projects

* [signalr-client](https://www.npmjs.com/package/signalr-client)

My work is inspired by _signalr-client_. Unfortunately, developer of _signalr-client_ is not working actively on it anymore.
Also, the way disconnection was managed in _signalr-client_ didn't suit my needs

* [node.bittrex.api](https://github.com/dparlevliet/node.bittrex.api)

_node.bittrex.api_ is a really nice wrapper around Bittrex API. Unfortunately it uses _signalr-client_ internally.

I need to add that without the work of [dparlevliet](https://github.com/dparlevliet) who did some reverse engineering on Bittrex usage of SignalR, this library would not exist


CHANGELOG.md
# Change Log

## [v1.1.10]
* New constructor option _resyncOrderBooksAutomatically_ can be set to _false_ to disable automatic order books resync (default = true). This might be useful when subscribing to many pairs
* New constructor option _exitOnQueryExchangeStateError_ can be set to _true_ to force library to exit in case an error occurs while trying to resync an order book (mostly for troubleshooting purpose) (default = false)

## [v1.1.9]
* New constructor option _markNewOrderBookEntriesAsUpdates_ can be set to _false_ to have _action_ = _add_ instead of _update_ when a new order book entries are added (default = _true_)

## [v1.1.8]
* An error was triggered when calling QueryExchangeState for an unsupported pair

## [v1.1.7]
* Fix previous version in case socket is reconnected

## [v1.1.6]
* Only return trades which were executed after subscription (this means that first 'trades' event will be emitted only when the first trade will be executed after subscription)

## [v1.1.5]
* Fix _ticker_ timestamp

## [v1.1.4]
* Add _trade id_ when emitting _trades_

## [v1.1.3]
* Only emit _watchdog_ event for _orders_ if authentication succeeded

## [v1.1.2]
* Ensures connection to exchange is always closed when we don't have any subscription remaining
* New constructor option _watchdog.orders_ to automatically force re-subscriptions for orders periodically
* Fix method _unsubscribeFromOrders_ (subscription wasn't cancelled)

## [v1.1.1]
* Allows to force re-subscription to orders (Bittrex Beta API)
* New method to enable logging keepalive messages (node _DEBUG_ must be enabled)
* Bittrex Beta is over, removed _legacy_ methods

## [v1.1.0]
* Support for Bittrex Beta API (tickers, order books, trades & user orders)
* Possibility to disable Cloud Scraper

## [v1.0.9]
* Change default User-Agent & add extra headers to bypass CloudFare protection

## [v1.0.8]
* Watchdog was added to reconnect upon detecting timeout (ie: when _Bittrex_ stopped sending data)

## [v1.0.7]
* Method _subscribeToAllTickers_ was added to subscribe to all tickers at once

## [v1.0.6]
* Take UTC offset into account when parsing DateTime strings returned by Bittrex

## [v1.0.5]
* Use _Big.js_ for floating point arithmetic

## [v1.0.4]
* Change _transport_ from _serverSentEvents_ to _webSockets_ for _abort_ step

## [v1.0.3]
* Change _transport_ from _serverSentEvents_ to _webSockets_ for _start_ step
* Implement call to _SignalR_ method _SubscribeToSummaryDeltas_ for tickers subscription (update are not sent automatically anymore by Bittrex)

## [v1.0.2]
* Ensure _disconnected_ and _connected_ events are properly emitted
* New method to retrieve SignalR connectionId
* Changes to logged messages


examples/subscribe_tickers.js
"use strict";
const util = require('util');
const SignalRClient = require('../lib/client.js');
let client = new SignalRClient({
    // websocket will be automatically reconnected if server does not respond to ping after 10s
    pingTimeout:10000
});

//-- event handlers
client.on('tickers', function(data){
    console.log(util.format("Got ticker update for pair '%s'", data.pair));
});

//-- start subscription
console.log("Subscribing to 'USDT-BTC' pair");
client.subscribeTickers(['USDT-BTC']);

// add subscription for 'USDT-ETH' & 'BTC-USDT' after 10s
setTimeout(function(){
    console.log("Adding subscription for USDT-ETH & BTC-ETH pairs");
    client.subscribeTickers(['USDT-ETH','BTC-ETH']);
}, 15000);

// add subscription for 'BTC-NEO', unsubscribe from previous pairs after 20s
setTimeout(function(){
    console.log("Setting BTC-NEO as the only pair we want to subscribe to");
    client.subscribeTickers(['BTC-NEO'], {reset:true});
}, 30000);

// disconnect client after 45s
setTimeout(function(){
    console.log('Disconnecting...');
    client.disconnect();
    process.exit(0);
}, 45000);


examples/subscribe_to_orders.js
"use strict";
const util = require('util');
const SignalRClient = require('../lib/client.js');

let client = new SignalRClient({
    // websocket will be automatically reconnected if server does not respond to ping after 10s
    pingTimeout:10000,
    // NB: you need to provide correct API key & secret above to be able to subscribe to orders
    auth:{
        key:"abcdef",
        secret: "123456"
    },
    watchdog:{
        // automatically re-subscribe for orders every 30min (this is enabled by default)
        orders:{
            enabled:true,
            period:1800
        }
    },
    // use cloud scraper to bypass Cloud Fare (default)
    useCloudScraper:true
});

//-- event handlers
client.on('order', function(data){
    console.log(util.format("Got 'order' event for order '%s' (%s)", data.orderNumber, data.pair));
    console.log(JSON.stringify(data));
});

//-- start subscription
console.log("=== Subscribing to orders");
client.subscribeToOrders();

// disconnect client after 10min
setTimeout(function(){
    console.log('=== Disconnecting...');
    client.disconnect();
    process.exit(0);
}, 600000);


examples/subscribe_to_tickers.js
"use strict";
const util = require('util');
const SignalRClient = require('../lib/client.js');
let client = new SignalRClient({
    // websocket will be automatically reconnected if server does not respond to ping after 10s
    pingTimeout:10000,
    // use cloud scraper to bypass Cloud Fare (default)
    useCloudScraper:true
});

//-- event handlers
client.on('ticker', function(data){
    console.log(util.format("Got ticker update for pair '%s'", data.pair));
});

//-- start subscription
console.log("=== Subscribing to 'USDT-BTC' pair");
client.subscribeToTickers(['USDT-BTC']);

// add subscription for 'USDT-ETH' & 'BTC-USDT' after 15s
setTimeout(function(){
    console.log("=== Adding subscription for USDT-ETH & BTC-ETH pairs");
    client.subscribeToTickers(['USDT-ETH','BTC-ETH']);
}, 30000);

// add subscription for 'BTC-NEO', unsubscribe from previous pairs after 30s
setTimeout(function(){
    console.log("=== Setting BTC-NEO as the only pair we want to subscribe to");
    client.subscribeToTickers(['BTC-NEO'], true);
}, 60000);

// disconnect client after 60s
setTimeout(function(){
    console.log('=== Disconnecting...');
    client.disconnect();
    process.exit(0);
}, 120000);


examples/subscribe_orderBooksAndTrades.js
"use strict";
const util = require('util');
const SignalRClient = require('../lib/client.js');
let client = new SignalRClient({
    // websocket will be automatically reconnected if server does not respond to ping after 10s
    pingTimeout:10000
});

//-- event handlers
client.on('orderBookAndTrades', function(data){
    console.log(util.format("Got full order book and trades for pair '%s' : cseq = %d", data.pair, data.cseq));
});
client.on('orderBookAndTradesUpdate', function(data){
    console.log(util.format("Got order book and trades update for pair '%s' : cseq = %d", data.pair, data.cseq));
});

//-- start subscription
console.log("Subscribing to 'USDT-BTC' pair");
client.subscribeOrderBooksAndTrades(['USDT-BTC']);

// add subscription for 'USDT-ETH' & 'BTC-USDT' after 10s
setTimeout(function(){
    console.log("Adding subscription for USDT-ETH & BTC-ETH pairs");
    client.subscribeOrderBooksAndTrades(['USDT-ETH','BTC-ETH']);
}, 15000);

// add subscription for 'BTC-NEO', unsubscribe from previous pairs after 20s
setTimeout(function(){
    console.log("Setting BTC-NEO as the only pair we want to subscribe to");
    client.subscribeOrderBooksAndTrades(['BTC-NEO'], {reset:true});
}, 30000);

// disconnect client after 45s
setTimeout(function(){
    console.log('Disconnecting...');
    client.disconnect();
    process.exit(0);
}, 45000);


examples/subscribe_to_markets.js
"use strict";
const util = require('util');
const SignalRClient = require('../lib/client.js');
let client = new SignalRClient({
    // websocket will be automatically reconnected if server does not respond to ping after 10s
    pingTimeout:10000,
    watchdog:{
        // automatically reconnect if we don't receive markets data for 30min (this is the default)
        markets:{
            timeout:1800000,
            reconnect:true
        }
    },
    // use cloud scraper to bypass Cloud Fare (default)
    useCloudScraper:true
});

//-- event handlers
client.on('orderBook', function(data){
    console.log(util.format("Got full order book for pair '%s' : cseq = %d", data.pair, data.cseq));
});
client.on('orderBookUpdate', function(data){
    console.log(util.format("Got order book update for pair '%s' : cseq = %d", data.pair, data.cseq));
});
client.on('trades', function(data){
    console.log(util.format("Got trades for pair '%s'", data.pair));
});

//-- start subscription
console.log("=== Subscribing to 'USDT-BTC' pair");
client.subscribeToMarkets(['USDT-BTC']);

// add subscription for 'USDT-ETH' & 'BTC-USDT' after 15s
setTimeout(function(){
    console.log("=== Adding subscription for USDT-ETH & BTC-ETH pairs");
    client.subscribeToMarkets(['USDT-ETH','BTC-ETH']);
}, 30000);

// add subscription for 'BTC-NEO', unsubscribe from previous pairs after 30s
setTimeout(function(){
    console.log("=== Setting BTC-NEO as the only pair we want to subscribe to");
    client.subscribeToMarkets(['BTC-NEO'], true);
}, 60000);

// disconnect client after 60s
setTimeout(function(){
    console.log('=== Disconnecting...');
    client.disconnect();
    process.exit(0);
}, 120000);
